### ISSUE TITLE
Investigate memory leaks when using subscriptions

### ISSUE THREAD

**Issue by nirinchev** [realm/realm-dotnet#2901]
A user sent me this:

> Hi Nikola,
we have most probably found the cause of the leaks. And now we need your help to find a way to fix it.
First of all some details to better describe how we use REALM DB.
>
> The app is a IoT SW that needs to be constantly up to date via Subscription to DB classes.
This app is up and running 24x7x365.
It’s never restarted or killed.
>
> Now, we go deeper in our issue.
When there is an update in one of the DB fileds, we see an increase of the app’s memory.
Of course this behaviour can be seen after many hours.
What have we done then to prove it:
>
> We commented only the write operation leaving the app running for many hours – we saw almost no memory increase
> We made a demo app ([GitHub - kalpasrl/RealmTest 2](https://github.com/kalpasrl/RealmTest)) to be run in the same conditions of our app
> We did a double test:
> Test 1 - With subscriptions and doing many write a minute - it’s a stress test in order to have results in a short time
> Test 2 - Without subscriptions doing the same test above
> Results:
> colors legenda:
> Brown line: native memory
> Orange line: private/managed memory
> Blue line: total memory
>
> Test 1 – the memory usage is shown in the following picture
> 
> ![](https://www.mongodb.com/community/forums/uploads/default/original/3X/f/e/fe9fab749bc9fe228ef1a21699f5aa0e7f7dd03a.jpeg)
> Test 2 – the memory usage is shown in the following picture
> ![](https://www.mongodb.com/community/forums/uploads/default/original/3X/1/b/1b2781c4f624acceeb2f62f9836b8de85d396d54.jpeg)
> As we can see, it seems that as soon as I activate subscriptions and start updating the native memory starts leaking.
> This behaviour is not seen if I do the test with subscriptions deactivated.
>
> The demo app is setup with subscriptions enabled.
> If you you want to run it without subscriptions, you have to comment the line 63 of App.xaml.cs file.

We should try to repro and understand what's causing this behavior.

**Comment by kalpasrl** [comment]
Hi @papafe,
the graphs shown are produced with our internal tool that collects the results from the adb command:
`adb shell dumpsys meminfo com.companyname.realmtest`

Best regards
Stefano

**Comment by papafe** [comment]
@kalpasrl Thanks for clarifying it!

**Comment by papafe** [comment]
@kalpasrl Can I ask you under which condition have you tested? 
In particular:
- Are you using a simulator or a real device?
- Are you using a debug or release build?
- Are you keeping the debugger attached while testing?

**Comment by papafe** [comment]
@kalpasrl Thanks for the info. I was trying to reproduce the issue with the project you linked, but I managed only when using the app compiled in debug mode, and with the debugger attached. 
I didn't manage to reproduce it in release mode, either with the debugger attached or not. 
Interestingly enough, when I managed to reproduce it (debug mode, debugger attached) I had a similar graph even by just using `Debug.WriteLine` with a fixed string (for example `Debug.WriteLine("something")`, instead of printing the number of indices in `changes`
